### ISSUE TITLE
master is not releasing memory

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#268]
I've taken several snapshots and master's RAM allocation grew to 11 GiB.
Then I removed some heavy snapshots (each accounts for about 1 GiB of RAM) but master's RAM allocation remained at maximum level.

**Comment by DarkHaze** [comment]
Hi Dmitry,

We are investigating the issue. We have just committed the patch with use of tcmalloc. It's fast and efficient memory allocation manager and it might help with the problem.
Also we are looking at the memory management in mfsmaster and possibly we will issue a patch soon.

**Comment by onlyjob** [comment]
Hi @DarkHaze, 

I just hope you've chosen something that we already have Debian. Are you talking about _gperftools_?

  https://packages.qa.debian.org/g/google-perftools.html
  https://code.google.com/p/gperftools/

**Comment by DarkHaze** [comment]
Hi Dmitry,

On debian required package name is libtcmalloc_minimal* ( last part vary
depending on debian release ) which is build from source package gperftools
( or google-perftools ).
So basically answer to your question is yes :).
But we also have found a situation in master where it keeps unused/released
data ( inodes, etc. )  for later use ( for up to 24h ) which could cause
your problem.
With advances in memory allocators this code doesn't make sense any more
and it will be removed soon.

**Comment by onlyjob** [comment]
Master certainly do not release memory even after 24h. It is not a big problem but it would be nice to fix if you could. Thanks.

**Comment by onlyjob** [comment]
According to #316 _tcmalloc_ is responsible for performance degradation in 3.9.2.
I'd rather have master that is not releasing memory than unresponsive one.
Please do not build with _tcmalloc_ by default.

**Comment by onlyjob** [comment]
[jemalloc](http://www.canonware.com/jemalloc/) might be a better alternative than _tcmalloc_:

 * https://stackoverflow.com/questions/7852731/c-memory-allocation-mechanism-performance-comparison-tcmalloc-vs-jemalloc
 * https://stackoverflow.com/questions/9866145/whatre-the-differences-between-tcmalloc-jemalloc-and-memory-pool

**Comment by krissfr** [comment]
any news on this issue ?

**Comment by psarna** [comment]
No news, but I can provide more context:
* It's not the case of master not releasing memory - master releases it, but its allocator (_malloc_ by default) keeps all pages allocated for performance reasons. Calling _free_ does not guarantee that acquired pages are released right away. In fact, it is almost always more effective to keep the pages - they would probably be needed soon.
* It is possible to override allocation routines with the help of external libraries, _tcmalloc_, _jemalloc_, etc. Most of them can be tried with LD_PRELOAD (http://stackoverflow.com/questions/426230/what-is-the-ld-preload-trick), but there is no universally good solution. Some libraries work good with certain use cases and terribly bad with others.
* Even if _malloc_ keeps redundant memory, it is not necessarily very bad. Most likely, if other processes need memory, system could just swap some of _malloc_'s unused pages (no performance drop here, since no one uses them anyway). If master works on a dedicated machine (which is advised), it should actually be encouraged to take as much memory as it can, if it could help him gain better performance.

tl;dr: For specific use cases, alternative allocation libraries can be considered useful, but it depends on several factors and should be preceded by performance tests. By default, LizardFS relies on _malloc_ and it will not be changed in near future.

**Comment by psarna** [comment]
Closing due to inactivity. Since we continue to use malloc, which usually doesn't free memory, but keeps it for later use anyway, the situation will probably stay as is.
